BSDT row service status active filter

AndersonTarren opened the pull request:
# Critical Changes

# Changes
Allows Service Status and Program Engagement Stages on BSDT rows to be filtered by values considered active.  This is configurable via custom metadata like our other bucketed configurations.

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [ ] Any net new LWC work has Sa11y tests & 50% or above lines JEST test coverage  
- [ ] CRUD/FLS is enforced in Apex
- [ ] Permission sets are updated to account for CRUD|FLS|Tab|Classes
- [ ] Field sets are updated to account for new fields
- [ ] UX approval or UX not necessary
- [x] Link the pull request and work item by PR comment and Chatter post respectively, e.g. GUS: W-0000000: Work Name (https://github.com/SalesforceFoundation/PMM/pull/303)
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [x] PR contains draft release notes
- [ ] QE story level testing completed

AndersonTarren commented:
[W-12236668](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001HfUrpYAF/view)

AndersonTarren commented:
[W-12236670](https://gus.lightning.force.com/a07EE00001HfUrrYAF)

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -11,19 +11,60 @@ public with sharing class ServiceService {
+    private Set<String> activeStatuses {
+        get {
+            if (activeStatuses == null) {
+                activeStatuses = getactiveStatuses();
```
typo, should be getActiveStatuses()

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -11,19 +11,60 @@ public with sharing class ServiceService {
+    }
+
+    private Set<String> getActiveStatuses() {
+        List<String> bucketNames = new List<String>{ 'ServiceStatusActive' };
```
Can we put this in a constant?

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -34,13 +75,18 @@ public with sharing class ServiceService {
-                .add(convertObjectToOption(engagement, engageProgram));
+            if (activeProgramEngagementStages.contains(engagement.Stage__c)) {
+                programIds.add(engagement.Program__c);
+                result.get('engagements')
```
This predates this PR but noting that we prolly should at least have these as final variables at the top of the method.

jjbennett commented on force-app/main/default/customMetadata/Bucket.ServiceStatusActive.md-meta.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<CustomMetadata xmlns="http://soap.sforce.com/2006/04/metadata" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+    <label>Service Status: Active</label>
```
IIRC the last time we added a filter we did not include any bucketed values and then in the code said if no bucketed values then include everything. This way the customer could still opt in. 
The concern is if an existing customer is not using the status field or representing Active with Active, it'd be a breaking change.

jjbennett reviewed (commented):
Tarren this looks great, I did have a concern and some minor suggestions.

jjbennett commented:
Also to note this feature might be a little confusing when a program engagement is active but not the service.
Using existing program engagement active buckets could also be a breaking change for some customers but not sure how we'd get them to opt in since those were already included.

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -11,19 +11,60 @@ public with sharing class ServiceService {
     private ServiceSelector serviceSelector = new ServiceSelector();
 
+    @TestVisible
+    private ProgramEngagementService engagementService = new ProgramEngagementService();
```
Not suggesting we change it but curious if it would be better to only include the service vs the service and the selector since the service calls the selector as well.

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -34,13 +75,18 @@ public with sharing class ServiceService {
-            programIds.add(engagement.Program__c);
-            result.get('engagements')
-                .add(convertObjectToOption(engagement, engageProgram));
+            if (activeProgramEngagementStages.contains(engagement.Stage__c)) {
```
Since customers can type in values, do we want to be case insensitive. I think historically we haven't been.

AndersonTarren commented:
Thanks for the great review @jjbennett - I'll get going on these suggestions.

AndersonTarren commented:
@jjbennett See what you think of these changes.  I corrected the typo and updated the constants.  Also implemented the change to use all values when no active items are defined.

That's a good question on case sensitivity - I think if we make a call on that, I'd like a focused WI for that where we investigate the entire bucket system to make sure it is aligned.

jjbennett commented:
@AndersonTarren We already included the metadata for active Program Engagements, this will automatically filter existing users BSDTs. Do we want to allow them to opt in somehow or do we want to do a critical change in the release notes? Then we can also note they can add a Service Status bucket value to filter that as well?

jjbennett commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -68,12 +69,28 @@ public with sharing class ServiceService {
         return activeStatuses;
     }
 
+    private Boolean getActiveFilterFeatureGateStatus() {
```
could we rename this method to something like isBsdtActiveFilterActive?

jjbennett reviewed (commented):
Nice! I haven't tossed this into an org again yet just had a couple comments after looking at the code again. Also, did we decide to add back the Active status for ServiceStatusActive bucket?

AndersonTarren commented on force-app/main/default/classes/ServiceService.cls:
```diff
@@ -68,12 +69,28 @@ public with sharing class ServiceService {
+
+        for (FeatureGate__mdt feature : features) {
+            if (
+                feature.IsActive__c && BSDT_ACTIVE_FILTER.contains(feature.DeveloperName)
```
I had similar thoughts, but I noticed in another location we're using this pattern.  Happy to move it to equals.
